## 폐암 환자의 생존율 예측하기 기본 예제 본인이 직접 실습해보기.

In [24]:
# 먼저 필요한 라이브러리를 불러옵니다,.

import tensorflow as tf
import numpy as np

In [25]:
# 딥러닝을 구동하는 데 필요한 케라스 함수를 호출합니다.

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [26]:
# 실행할때마다 같은 결과를 출력하기 위해 설정하는 부분

np.random.seed(3)
tf.random.set_seed(3)

In [27]:
# 준비된 수술 환자 데이터를 불러오기. (케글에서의 CSV 파일.)

Data_set = np.loadtxt("C:/Users/korea/Desktop/My Jupyter/Thoraric Surgery Patient/ThoraricSurgery.csv", delimiter=",")

# 경로를 정확히 지정하고, delimiter는 무엇인가?
# 구분 문자는 일반 텍스트 또는 데이터 스트림에서 별도의 독립적 영역 사이의 경계를 지정하는 데 사용하는 
# 하나의 문자 혹은 문자들의 배열을 말한다. 구분 문자의 한 예로 쉼표가 있는데, CSV 파일의 필드 구분자의 역할을 한다. 
# 영어로 Delimit이 "범위를 정하다"라는 의미이다. 

In [28]:
# 환자의 기록과 수술 결과를 X와 Y로 구분하여 저장

X = Data_set[:,0:17]
Y = Data_set[:,17]

In [29]:
# 그 다음, 딥러닝 구조를 결정(모델을 설정하고 실행)

model = Sequential()
model.add(Dense(30, input_dim=17, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [30]:
# 딥러닝을 실제로 실행해보자.

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X, Y, epochs=100, batch_size=10)

Epoch 1/100
47/47 [==============================] - 0s 476us/step - loss: 0.6482 - accuracy: 0.8128
Epoch 2/100
47/47 [==============================] - 0s 455us/step - loss: 0.4890 - accuracy: 0.8468
Epoch 3/100
47/47 [==============================] - 0s 434us/step - loss: 0.4416 - accuracy: 0.8511
Epoch 4/100
47/47 [==============================] - 0s 434us/step - loss: 0.4863 - accuracy: 0.8489
Epoch 5/100
47/47 [==============================] - 0s 434us/step - loss: 0.4430 - accuracy: 0.8532
Epoch 6/100
47/47 [==============================] - 0s 455us/step - loss: 0.4303 - accuracy: 0.8532
Epoch 7/100
47/47 [==============================] - 0s 455us/step - loss: 0.4421 - accuracy: 0.8511
Epoch 8/100
47/47 [==============================] - 0s 434us/step - loss: 0.4363 - accuracy: 0.8489
Epoch 9/100
47/47 [==============================] - 0s 455us/step - loss: 0.4165 - accuracy: 0.8489
Epoch 10/100
47/47 [==============================] - 0s 434us/step - loss: 0.4317 - accura

47/47 [==============================] - 0s 455us/step - loss: 0.4057 - accuracy: 0.8532
Epoch 82/100
47/47 [==============================] - 0s 477us/step - loss: 0.4184 - accuracy: 0.8489
Epoch 83/100
47/47 [==============================] - 0s 434us/step - loss: 0.3944 - accuracy: 0.8532
Epoch 84/100
47/47 [==============================] - 0s 477us/step - loss: 0.4345 - accuracy: 0.8468
Epoch 85/100
47/47 [==============================] - 0s 520us/step - loss: 0.4050 - accuracy: 0.8489
Epoch 86/100
47/47 [==============================] - 0s 455us/step - loss: 0.3922 - accuracy: 0.8489
Epoch 87/100
47/47 [==============================] - 0s 455us/step - loss: 0.3976 - accuracy: 0.8511
Epoch 88/100
47/47 [==============================] - 0s 520us/step - loss: 0.3966 - accuracy: 0.8468
Epoch 89/100
47/47 [==============================] - 0s 499us/step - loss: 0.3886 - accuracy: 0.8468
Epoch 90/100
47/47 [==============================] - 0s 737us/step - loss: 0.3965 - accuracy: 

# 여기서 주목해야 할 것은, 마지막 accuracy(정확도)와 손실 값(loss) 이다.
* loss는 예측에서 빗나간 수치, accuracy는 예측이 성공할 확률이다. 만약 accuracy 수치가 1.0에 근접한다면 확률이 100%에 근접하다는 의미.
지금 정확도 수치는 마지막의 0.8468 수치를 참고하면 84.68%가 된다는 것이다.

# 이러한 간단한 코드로 실행되는 딥러닝 구조 자체는 단순하지만, 결과는 놀랍다. 하지만, 데이터를 분석해 데이터를 확장하고,
# 딥러닝 구조를 적절하게 바꾸는 등의 노력으로 더 향상될 수 있다. 